### PR TITLE
New subcommand for agama logs, refactoring

### PR DESCRIPTION
**Pull request by mchf**
## Problem

Trying to grow agama logs from baby age to teenage.

## Solution

- Implemented agama logs list subcommand
- Internal refactoring
- Changes in target tarball - related to: https://github.com/openSUSE/agama/issues/774
- Commands are not stored in files with names containing spaces anymore. 
- Command generated logs can be stored in files with user defined names

## Testing

- *Tested manually*

### CODE REVIEW

File: rust/agama-cli/src/logs.rs [hunk 1]
```diff
@@ -22,10 +23,25 @@ pub enum LogsCommands {
...
 
+// main entry point called from agama CLI main loop
 pub async fn run(subcommand: LogsCommands) -> anyhow::Result<()> {
     match subcommand {
-        LogsCommands::Store { verbose } => Ok(store(verbose)?),
-        LogsCommands::List => Err(anyhow::anyhow!("Not implemented")),
+        LogsCommands::Store { verbose } => {
+            // feed internal options structure by what was received from user
+            // for now we always use / add defaults if any
+            let mut options = LogOptions::new();
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:32]: I would suggest to implement [Default](https://doc.rust-lang.org/std/default/trait.Default.html) for `LogOptions` and change this code to:

```rust
let options = LogOptions { verbose, ..Default::default() }
```

And the same below (see the following comment).

File: rust/agama-cli/src/logs.rs [hunk 2]
```diff
@@ -22,10 +23,25 @@ pub enum LogsCommands {
...
+
+            options.verbose = verbose;
+
+            Ok(store(options)?)
+        }
+        LogsCommands::List => {
+            let options = LogOptions::new();
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:39]: You can change this to (if Default is implemented):

```rust
            list(LogOptions::default());
            Ok(())
```

File: rust/agama-cli/src/logs.rs [hunk 3]
```diff
@@ -76,6 +92,24 @@ fn show(show: bool, text: &str) {
...
 
+// Configurable parameters of the "agama logs" which can be
+// set by user when calling a (sub)command
+struct LogOptions {
+    paths: Vec<String>,
+    commands: Vec<String>,
+    verbose: bool,
+}
+
+impl LogOptions {
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:103]:
```rust
impl Default for LogOptions {
    fn default() -> Self {
        Self {
            paths: DEFAULT_PATHS.iter().map(ToString::to_string).collect(),
            commands: DEFAULT_COMMANDS.iter().map(ToString::to_string).collect(),
            verbose: false,
        }
    }
}
```

File: rust/agama-cli/src/logs.rs [hunk 4]
```diff
@@ -181,64 +218,77 @@ impl LogItem for LogCmd {
...
 }
 
-// compress given directory into a tar archive
+// Compress given directory into a tar archive
 fn compress_logs(tmp_dir: &TempDir, result: &String) -> io::Result<()> {
     let compression = DEFAULT_COMPRESSION.0;
+    let tmp_path = tmp_dir
+        .path()
+        .parent()
+        .unwrap()
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:253]: It would be nice to get rid of this `unwrap()` :wink:

File: rust/agama-cli/src/logs.rs [hunk 5]
```diff
@@ -181,64 +218,77 @@ impl LogItem for LogCmd {
...
+        .as_os_str()
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:254]: Why do you need to call `as_os_str()` here?

File: rust/agama-cli/src/logs.rs [hunk 6]
```diff
@@ -181,64 +218,77 @@ impl LogItem for LogCmd {
...
+        .to_str()
+        .ok_or(io::Error::new(io::ErrorKind::InvalidInput, "Wrong path"))?;
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:256]: Adding the path would be useful.

File: rust/agama-cli/src/logs.rs [hunk 7]
```diff
@@ -181,64 +218,77 @@ impl LogItem for LogCmd {
...
-        .status()
-    {
-        Ok(_o) => Ok(()),
-        Err(_e) => Err(io::Error::new(
+        .status()?;
+
+    if res.success() {
+        Ok(())
+    } else {
+        Err(io::Error::new(
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:275]: Perhaps you could define your own error type instead of reusing the `io::Error`.

File: rust/agama-cli/src/logs.rs [hunk 8]
```diff
@@ -278,3 +328,14 @@ fn store(verbose: bool) -> Result<(), io::Error> {
 
     compress_logs(&tmp_dir, &result)
 }
+
+// Handler for the "agama logs list" subcommand
+fn list(options: LogOptions) {
+    for list in [options.paths, options.commands] {
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:334]: It would indicate what is a file and what is a command in the UI.

**Review by imobachgs — COMMENTED** [review]
In general, it looks good. Just a few suggestions.

File: rust/agama-cli/src/logs.rs (hunk 1 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:32]: ah ... another chapter ... nice tip ;-)

File: rust/agama-cli/src/logs.rs (hunk 2 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:39]: done

File: rust/agama-cli/src/logs.rs (hunk 3 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:103]: already implemented as a consequence of previous sugestions ;-)

File: rust/agama-cli/src/logs.rs (hunk 5 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:254]: I need to convert it somehow into a list of characters. As far as I understand the implementation, if you used e.g. display it may remove some weird (unprintable) characters from path. This way it should convert as much as possible. I consider such case a little bit obscure thought  but better safe than sorry ;-)

File: rust/agama-cli/src/logs.rs (hunk 6 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:256]: Well ... it would ... But conversion of tmp dir's ```Path``` into a usable string is basically what fails here, so I cannot print it. May be I can rephrase it to something like "Temporary directory cannot be found"

File: rust/agama-cli/src/logs.rs (hunk 8 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:334]: slightly modified

File: rust/agama-cli/src/logs.rs (hunk 5 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:254]: And if you asked why I need ```as_os_str``` here and not in the subsequent case that's because here ```parent``` returns ```Path``` but ```file_name``` used later returns ```OsStr``` already.

File: rust/agama-cli/src/logs.rs (hunk 4 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:253]: tried so

File: rust/agama-cli/src/logs.rs (hunk 7 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:275]: Well ... perhaps ... but I somehow feel it as an io error here ;-)

File: rust/agama-cli/src/logs.rs (hunk 5 above)
Comment (imobachgs) [reply rust/agama-cli/src/logs.rs:256]: To be honest, I would not expect obscure characters here. But we can keep it as it is.

File: rust/agama-cli/src/logs.rs (hunk 7 above)
Comment (imobachgs) [reply rust/agama-cli/src/logs.rs:286]: OK, that's fine. I usually prefer to have my error types (keeping the source of the error). But given that `io::Error` belongs to the standard library, let's keep it as it is.

**Review by imobachgs — COMMENTED** [review]
In general, it looks better. However, I still think that the content of the tarball looks kind of weird:

```
agama-logs.UrXFYtb6nDY7/journalctl--dmesg.err.log
agama-logs.UrXFYtb6nDY7/journalctl--dmesg.out.log
agama-logs.UrXFYtb6nDY7/journalctl-uagama-auto.err.log
agama-logs.UrXFYtb6nDY7/journalctl-uagama-auto.out.log
agama-logs.UrXFYtb6nDY7/journalctl-uagama.err.log
agama-logs.UrXFYtb6nDY7/journalctl-uagama.out.log
```


IMHO, it would be better to have something like `agama.log`, `agama.err`,  `dmesg.log`, `dmesg.err`, `agama-auto.log`, etc. Actually, I am not even sure whether I would keep the `.err` files, but that's another discussion.

Instead of inferring the name from the command, I think extending the `LogCmd` struct to include a basename (or some similar mechanism) is better.

File: rust/agama-cli/src/logs.rs [hunk 9]
```diff
@@ -181,64 +230,96 @@ impl LogItem for LogCmd {
...
+        .ok_or(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Malformed path to temporary directory",
+        ))?
+        .as_os_str()
+        .to_str()
+        .ok_or(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Malformed path to temporary directory",
+        ))?;
```
Comment (imobachgs) [inline rust/agama-cli/src/logs.rs:281]: Sorry, but I see a lot of repetition here with four errors that seem to be the same (but they are not). Anyway, if you are not interested in making any distinction, you could rewrite this code to:

```rust
    let path = tmp_dir.path();
    let tmp_path = path
        .parent()
        .map(|p| p.as_os_str().to_str())
        .flatten()
        .ok_or(io::Error::new(
            io::ErrorKind::InvalidInput,
            "Malformed path to temporary directory",
        ))?;
    let dir = path
        .file_name()
        .map(|p| p.to_str())
        .flatten()
        .ok_or(io::Error::new(
            io::ErrorKind::InvalidInput,
            "Malformed path to temporary directory",
        ))?;
```

And I still see some repetition... :-)

**Review by imobachgs — COMMENTED** [review]
It looks better, but there is still some repetition.

File: rust/agama-cli/src/logs.rs [hunk 10]
```diff
@@ -22,17 +23,31 @@ pub enum LogsCommands {
...
+        }
     }
 }
 
-const DEFAULT_COMMANDS: [&str; 3] = [
-    "journalctl -u agama",
-    "journalctl -u agama-auto",
-    "journalctl --dmesg",
+const DEFAULT_COMMANDS: [(&str, &str); 3] = [
+    ("journalctl -u agama", "agama"),
```
[the remark this answers is not in the document]
Comment (mchf) [reply rust/agama-cli/src/logs.rs:48]: a few words written

File: rust/agama-cli/src/logs.rs [hunk 11]
```diff
@@ -76,6 +91,27 @@ fn show(show: bool, text: &str) {
...
+    verbose: bool,
+}
+
+impl Default for LogOptions {
+    fn default() -> Self {
+        Self {
+            paths: DEFAULT_PATHS.iter().map(|p| p.to_string()).collect(),
+            commands: DEFAULT_COMMANDS
+                .iter()
+                .map(|p| (p.0.to_string(), p.1.to_string()))
```
[the remark this answers is not in the document]
Comment (mchf) [reply rust/agama-cli/src/logs.rs:108]: why not. Done when touching the code anyway

File: rust/agama-cli/src/logs.rs [hunk 12]
```diff
@@ -22,17 +23,31 @@ pub enum LogsCommands {
...
+// main entry point called from agama CLI main loop
 pub async fn run(subcommand: LogsCommands) -> anyhow::Result<()> {
     match subcommand {
-        LogsCommands::Store { verbose } => Ok(store(verbose)?),
-        LogsCommands::List => Err(anyhow::anyhow!("Not implemented")),
+        LogsCommands::Store { verbose } => {
+            // feed internal options structure by what was received from user
+            // for now we always use / add defaults if any
+            let options = LogOptions {
+                verbose: verbose,
```
[the remark this answers is not in the document]
Comment (mchf) [reply rust/agama-cli/src/logs.rs:33]: well i don't like this. I like to be explicit even if it costs some more letters. But if you wish so ...

File: rust/agama-cli/src/logs.rs [hunk 13]
```diff
@@ -181,64 +230,96 @@ impl LogItem for LogCmd {
...
+        )));
     }
 
     log_sources
 }
 
-// compress given directory into a tar archive
+// Compress given directory into a tar archive
 fn compress_logs(tmp_dir: &TempDir, result: &String) -> io::Result<()> {
     let compression = DEFAULT_COMPRESSION.0;
```
[the remark this answers is not in the document]
Comment (mchf) [reply rust/agama-cli/src/logs.rs:268]: placed a few words close to the DEFAULT_COMPRESSION definition

File: rust/agama-cli/src/logs.rs (hunk 9 above)
Comment (mchf) [reply rust/agama-cli/src/logs.rs:281]: well I can write slightly different message into each error and there won't be any repetition anymore ;-)

Comment (mchf) [reply rust/agama-cli/src/logs.rs:284]: btw i slightly doubt that the snippet works ... bcs ```as_os_str().to_str()``` returns ```Option<&str>```, so you need to unpack it ... that's why I have another ```ok_or``` in my code.

Comment (imobachgs) [reply rust/agama-cli/src/logs.rs:284]: I tried the code before suggesting the change. [parent](https://doc.rust-lang.org/std/path/struct.Path.html#method.parent) already returns an Option, so I call [map](https://doc.rust-lang.org/std/option/enum.Option.html#method.map) to work with the `Some` case. As you said, `as_os_str().to_str()` returns another option, so you get an `Option<Option<&str>>`. And that's why I call [flatten](https://doc.rust-lang.org/std/option/enum.Option.html?search=flatten#method.flatten), which returns a single `Option<&str>` and allows me to call `ok_or`.

Comment (mchf) [reply rust/agama-cli/src/logs.rs:284]: well to be honest, I see the suggested code a little bit less readable. However I haven't found anything better so far.

Comment (mchf) [reply rust/agama-cli/src/logs.rs:284]: I mean ... it can be of course perfectly fine ... but i need to invest into it a bit to make it common to me ;-)

Comment (mchf) [reply rust/agama-cli/src/logs.rs:284]: on the end after enough reading i found something similar, but one word shorter ;-) So, I'll test it and push if it works ;-)

Comment (imobachgs) [reply rust/agama-cli/src/logs.rs:279]: Sure, and_then looks like a better option here. Actually, *Rust by example* [confirms It](https://doc.rust-lang.org/rust-by-example/error/option_unwrap/and_then.html).

**Review by imobachgs — APPROVED** [review]
